In [3]:
import csv

In [4]:
insurance_data = []
with open('insurance.csv') as insurance_csv:
    insurance_reader = csv.DictReader(insurance_csv, delimiter = ',')
    for row in insurance_reader:
        insurance_data.append(row)

In [ ]:
insurance_data.pop(-1)

In [13]:
import statistics

# Extract 'age' values from the list of dictionaries
ages = [int(person['age']) for person in insurance_data]

# Calculate the minimum, maximum, median, and average age
min_age = min(ages)
max_age = max(ages)
median_age = statistics.median(ages)
avg_age = round(sum(ages) / len(ages), 2)

print("Minimum age:", min_age)
print("Maximum age:", max_age)
print("Median age:", median_age)
print("Average age:", avg_age)

Minimum age: 18
Maximum age: 64
Median age: 39.0
Average age: 39.21


In [14]:
# Initialize counters for males and females
male_count = 0
female_count = 0

# Count the occurrences of males and females
for person in insurance_data:
    if person['sex'] == 'male':
        male_count += 1
    if person['sex'] == 'female':
        female_count += 1

# Calculating the ratio between males and females
if female_count == 0:
    ratio = 'infinity'
else:
    fm_ratio = round(male_count / female_count, 2)

print("Male count:", male_count)
print("Female count:", female_count)
print("Ratio (Males : Females):", fm_ratio)

Male count: 676
Female count: 662
Ratio (Males : Females): 1.02


In [15]:
# Extract 'bmi' values from the list of dictionaries
bmis = [float(person['bmi']) for person in insurance_data]

# Calculate the minimum, maximum, median, and average BMI
min_bmi = min(bmis)
max_bmi = max(bmis)
median_bmi = statistics.median(bmis)
avg_bmi = round(sum(bmis) / len(bmis), 2)

print("Minimum BMI:", min_bmi)
print("Maximum BMI:", max_bmi)
print("Median BMI:", median_bmi)
print("Average BMI:", avg_bmi)

Minimum BMI: 15.96
Maximum BMI: 53.13
Median BMI: 30.4
Average BMI: 30.66


In [16]:
# Initialize counts for people with 0 children and people with 1-5 children
count_0_children = 0
count_1_to_5_children = 0

# Iterate through the insurance data
for person in insurance_data:
    num_children = int(person['children'])
    if num_children == 0:
        count_0_children += 1
    elif 1 <= num_children <= 5:
        count_1_to_5_children += 1

# Calculate the ratio
if count_0_children > 0:
    ratio_1_to_5_to_0 = round(count_1_to_5_children / count_0_children, 2)
else:
    ratio_1_to_5_to_0 = 0

print("Ratio of people with 1-5 children to those with 0 children:", ratio_1_to_5_to_0)

Ratio of people with 1-5 children to those with 0 children: 1.33


In [17]:
# Initialize counters for smokers and non-smokers
smoker_count = 0
non_smoker_count = 0

# Count the occurrences of smokers and non-smokers
for person in insurance_data:
    if person['smoker'] == 'yes':
        smoker_count += 1
    if person['smoker'] == 'no':
        non_smoker_count += 1

# Calculating the ratio between Non-Smokers and Smokers
if non_smoker_count == 0:
    ratio = 'infinity'
else:
    sns_ratio = round(non_smoker_count / smoker_count, 2)

print("Smoker count:", smoker_count)
print("Non-Smoker count:", non_smoker_count)
print("Ratio (Non-Smoker : Smoker):", sns_ratio)

Smoker count: 274
Non-Smoker count: 1064
Ratio (Non-Smoker : Smoker): 3.88


In [18]:
# Initialize a dictionary to store the count of each region
region_counts = {}

# Iterate over the list of dictionaries
for person in insurance_data:
    # Get the region from the current dictionary
    region = person['region']
    
    # If the region is not already recorded, initialize its count to 1
    if region not in region_counts:
        region_counts[region] = 1
    # If the region is already recorded, increment its count
    else:
        region_counts[region] += 1

# Calculate the total count of all regions
total_count = sum(region_counts.values())

# Print the count and ratio of each region relative to the total count
for region, count in region_counts.items():
    ratio = count / total_count
    print(f"{region} : Count: {count}, Ratio: {ratio:.2f}")

southwest : Count: 325, Ratio: 0.24
southeast : Count: 364, Ratio: 0.27
northwest : Count: 325, Ratio: 0.24
northeast : Count: 324, Ratio: 0.24


In [28]:
# Extract 'charges' values from the list of dictionaries
charges = [float(person['charges']) for person in insurance_data]

# Calculate the minimum, maximum, average, and median charges
min_charges = round(min(charges), 2)
max_charges = round(max(charges), 2)
avg_charges = round(sum(charges) / len(charges), 2)
median_charges = round(statistics.median(charges), 2)

print("Minimum charges:", min_charges)
print("Maximum charges:", max_charges)
print("Average charges:", avg_charges)
print("Median charges:", median_charges)

Minimum charges: 1121.87
Maximum charges: 63770.43
Average charges: 13270.42
Median charges: 9382.03


In [20]:
# Initialize dictionaries to store charges and counts for each age range
charges_by_age_range = {'15-24': [], '25-34': [], '35-44': [], '45-54': [], '55-64': []}
counts_by_age_range = {'15-24': 0, '25-34': 0, '35-44': 0, '45-54': 0, '55-64': 0}

# Categorize individuals into their respective age ranges and record charges
for person in insurance_data:
    age = int(person['age'])
    charges = float(person['charges'])  # Convert charges to float
    
    if 15 <= age <= 24:
        charges_by_age_range['15-24'].append(charges)
        counts_by_age_range['15-24'] += 1
    elif 25 <= age <= 34:
        charges_by_age_range['25-34'].append(charges)
        counts_by_age_range['25-34'] += 1
    elif 35 <= age <= 44:
        charges_by_age_range['35-44'].append(charges)
        counts_by_age_range['35-44'] += 1
    elif 45 <= age <= 54:
        charges_by_age_range['45-54'].append(charges)
        counts_by_age_range['45-54'] += 1
    elif 55 <= age <= 64:
        charges_by_age_range['55-64'].append(charges)
        counts_by_age_range['55-64'] += 1

# Calculate the average charges for each age range
avg_charges_by_age_range = {}
for age_range, charges_list in charges_by_age_range.items():
    total_charges = sum(charges_list)
    count = counts_by_age_range[age_range]
    if count != 0:
        avg_charges_by_age_range[age_range] = round(total_charges / count, 2)
    else:
        avg_charges_by_age_range[age_range] = 0

# Print the average charges for each age range
for age_range, avg_charges in avg_charges_by_age_range.items():
    print("Average charges for age range", age_range, ":", avg_charges)

Average charges for age range 15-24 : 9011.34
Average charges for age range 25-34 : 10352.39
Average charges for age range 35-44 : 13134.17
Average charges for age range 45-54 : 15853.93
Average charges for age range 55-64 : 18513.28


In [21]:
# Initialize dictionaries to store charges and counts for each BMI range
charges_by_bmi_range = {'15-24': [], '25-34': [], '35-44': [], '45-54': []}
counts_by_bmi_range = {'15-24': 0, '25-34': 0, '35-44': 0, '45-54': 0}

# Categorize individuals into their respective BMI ranges and record charges
for person in insurance_data:
    bmi = float(person['bmi'])
    charges = float(person['charges'])
    
    if 15 <= bmi <= 24:
        charges_by_bmi_range['15-24'].append(charges)
        counts_by_bmi_range['15-24'] += 1
    elif 25 <= bmi <= 34:
        charges_by_bmi_range['25-34'].append(charges)
        counts_by_bmi_range['25-34'] += 1
    elif 35 <= bmi <= 44:
        charges_by_bmi_range['35-44'].append(charges)
        counts_by_bmi_range['35-44'] += 1
    elif 45 <= bmi <= 54:
        charges_by_bmi_range['45-54'].append(charges)
        counts_by_bmi_range['45-54'] += 1

# Calculate the average charges for each BMI range
avg_charges_by_bmi_range = {}
for bmi_range, charges_list in charges_by_bmi_range.items():
    total_charges = sum(charges_list)
    count = counts_by_bmi_range[bmi_range]
    if count != 0:
        avg_charges_by_bmi_range[bmi_range] = round(total_charges / count, 2)
    else:
        avg_charges_by_bmi_range[bmi_range] = 0

# Print the average charges for each BMI range
for bmi_range, avg_charges in avg_charges_by_bmi_range.items():
    print(f"Average charges for BMI range {bmi_range}: {avg_charges}")

Average charges for BMI range 15-24: 9598.32
Average charges for BMI range 25-34: 12402.73
Average charges for BMI range 35-44: 16994.67
Average charges for BMI range 45-54: 17547.93


In [22]:
# Initialize variables to store total charges for males and females
total_charges_male = 0
total_charges_female = 0

# Calculate total charges and counts for males and females
count_male = 0
count_female = 0
for person in insurance_data:
    charges = float(person['charges'])  # Convert charges to float
    if person['sex'] == 'male':
        total_charges_male += charges
        count_male += 1
    else:
        total_charges_female += charges
        count_female += 1

# Calculate average charges for males and females
avg_charges_male = round(total_charges_male / count_male, 2) if count_male != 0 else 0
avg_charges_female = round(total_charges_female / count_female, 2) if count_female != 0 else 0

# Calculate the average percentage increase in charges for males compared to females
avg_percentage_increase = ((avg_charges_male - avg_charges_female) / avg_charges_female) * 100

print("Average charges for males:", avg_charges_male)
print("Average charges for females:", avg_charges_female)
print(f"The average percentage increase in charges for males compared to females is: {avg_percentage_increase:.2f}%")

Average charges for males: 13956.75
Average charges for females: 12569.58
The average percentage increase in charges for males compared to females is: 11.04%


In [23]:
# Initialize dictionaries to store charges and counts for each number of children
charges_by_children = {'0': [], '1': [], '2': [], '3': [], '4': [], '5': []}
counts_by_children = {'0': 0, '1': 0, '2': 0, '3': 0, '4': 0, '5': 0}

# Categorize individuals into their respective number of children and record charges
for person in insurance_data:
    num_children = person['children']
    charges = float(person['charges'])
    
    if num_children == '0':
        charges_by_children['0'].append(charges)
        counts_by_children['0'] += 1
    elif num_children == '1':
        charges_by_children['1'].append(charges)
        counts_by_children['1'] += 1
    elif num_children == '2':
        charges_by_children['2'].append(charges)
        counts_by_children['2'] += 1
    elif num_children == '3':
        charges_by_children['3'].append(charges)
        counts_by_children['3'] += 1
    elif num_children == '4':
        charges_by_children['4'].append(charges)
        counts_by_children['4'] += 1
    elif num_children == '5':
        charges_by_children['5'].append(charges)
        counts_by_children['5'] += 1

# Calculate the average charges per child for each number of children
avg_charges_by_children = {}
for num_children, charges_list in charges_by_children.items():
    total_charges = sum(charges_list)
    count = counts_by_children[num_children]
    if count != 0:
        avg_charges_by_children[num_children] = round(total_charges / count, 2)
    else:
        avg_charges_by_children[num_children] = 0

# Print the average charges per child for each number of children
for num_children, avg_charges in avg_charges_by_children.items():
    print(f"Average charges per child for {num_children} children: {avg_charges}")

Average charges per child for 0 children: 12365.98
Average charges per child for 1 children: 12731.17
Average charges per child for 2 children: 15073.56
Average charges per child for 3 children: 15355.32
Average charges per child for 4 children: 13850.66
Average charges per child for 5 children: 8786.04


In [24]:
# Initialize dictionaries to store charges and counts for smokers and non-smokers
charges_by_smoker = {'yes': [], 'no': []}
counts_by_smoker = {'yes': 0, 'no': 0}

# Categorize individuals into smokers and non-smokers and record charges
for person in insurance_data:
    smoker = person['smoker']
    charges = float(person['charges'])
    
    charges_by_smoker[smoker].append(charges)
    counts_by_smoker[smoker] += 1

# Calculate the average charges for smokers and non-smokers
avg_charges_by_smoker = {}
for smoker, charges_list in charges_by_smoker.items():
    total_charges = sum(charges_list)
    count = counts_by_smoker[smoker]
    if count != 0:
        avg_charges_by_smoker[smoker] = round(total_charges / count, 2)
    else:
        avg_charges_by_smoker[smoker] = 0

# Print the average charges for smokers and non-smokers
for smoker, avg_charges in avg_charges_by_smoker.items():
    print(f"Average charges for {'smokers' if smoker == 'yes' else 'non-smokers'}: {avg_charges}")

# Calculate the average percentage increase in charges for smokers compared to non-smokers
percentage_increase_smokers = ((avg_charges_by_smoker['yes'] - avg_charges_by_smoker['no']) / avg_charges_by_smoker['no']) * 100

print(f"Average percentage increase in charges for smokers compared to non-smokers: {percentage_increase_smokers:.2f}%")

Average charges for smokers: 32050.23
Average charges for non-smokers: 8434.27
Average percentage increase in charges for smokers compared to non-smokers: 280.00%


In [25]:
# Initialize dictionaries to store charges and counts for each region
charges_by_region = {'southwest': [], 'southeast': [], 'northwest': [], 'northeast': []}
counts_by_region = {'southwest': 0, 'southeast': 0, 'northwest': 0, 'northeast': 0}

# Categorize individuals into their respective regions and record charges
for person in insurance_data:
    region = person['region']
    charges = float(person['charges'])
    
    charges_by_region[region].append(charges)
    counts_by_region[region] += 1

# Calculate the average charges for each region
avg_charges_by_region = {}
for region, charges_list in charges_by_region.items():
    total_charges = sum(charges_list)
    count = counts_by_region[region]
    if count != 0:
        avg_charges_by_region[region] = round(total_charges / count, 2)
    else:
        avg_charges_by_region[region] = 0

# Print the average charges for each region and the count of persons in each region
for region, avg_charges in avg_charges_by_region.items():
    print(f"Average charges for {region}: {avg_charges}, Count: {counts_by_region[region]}")

Average charges for southwest: 12346.94, Count: 325
Average charges for southeast: 14735.41, Count: 364
Average charges for northwest: 12417.58, Count: 325
Average charges for northeast: 13406.38, Count: 324


In [26]:
# Initialize dictionaries to store charges and counts for each group
charges_by_group = {
    'male_yes': [], 'male_no': [],
    'female_yes': [], 'female_no': []
}
counts_by_group = {
    'male_yes': 0, 'male_no': 0,
    'female_yes': 0, 'female_no': 0
}

# Categorize individuals into groups based on sex and smoking status and record charges
for person in insurance_data:
    sex = person['sex']
    smoker = person['smoker']
    charges = float(person['charges'])
    
    group = f"{sex}_{smoker}"
    charges_by_group[group].append(charges)
    counts_by_group[group] += 1

# Calculate the average charges for each group
avg_charges_by_group = {}
for group, charges_list in charges_by_group.items():
    total_charges = sum(charges_list)
    count = counts_by_group[group]
    if count != 0:
        avg_charges_by_group[group] = round(total_charges / count, 2)
    else:
        avg_charges_by_group[group] = 0

# Print the average charges for each group
for group, avg_charges in avg_charges_by_group.items():
    sex, smoker = group.split('_')
    print(f"Average charges for {sex} {'smokers' if smoker == 'yes' else 'non-smokers'}: {avg_charges}")

Average charges for male smokers: 33042.01
Average charges for male non-smokers: 8087.2
Average charges for female smokers: 30679.0
Average charges for female non-smokers: 8762.3


In [29]:
# Initialize dictionaries to store charges and counts for each group
charges_by_group = {}
counts_by_group = {}

# Categorize individuals into groups based on age range, sex, smoking status, and number of children
for person in insurance_data:
    age = int(person['age'])
    sex = person['sex']
    smoker = person['smoker']
    children = int(person['children'])
    charges = float(person['charges'])
    
    age_range = ''
    if 15 <= age <= 24:
        age_range = '15-24'
    elif 25 <= age <= 34:
        age_range = '25-34'
    elif 35 <= age <= 44:
        age_range = '35-44'
    elif 45 <= age <= 54:
        age_range = '45-54'
    elif 55 <= age <= 64:
        age_range = '55-64'
    
    group = f"{age_range}_{sex}_{smoker}_{children}"
    
    if group not in charges_by_group:
        charges_by_group[group] = []
        counts_by_group[group] = 0
    
    charges_by_group[group].append(charges)
    counts_by_group[group] += 1

# Calculate the average charges for each group
avg_charges_by_group = {}
for group, charges_list in charges_by_group.items():
    total_charges = sum(charges_list)
    count = counts_by_group[group]
    if count != 0:
        avg_charges_by_group[group] = round(total_charges / count, 2)
    else:
        avg_charges_by_group[group] = 0

# Organize groups by age range
age_ranges = ['15-24', '25-34', '35-44', '45-54', '55-64']

# Print the average charges for each group by age range, gender, smoking status, and number of children
for age_range in age_ranges:
    print(f"Average charges for age range {age_range}:")
    for sex in ['female', 'male']:
        for smoker in ['yes', 'no']:
            for children in sorted(set(int(group.split('_')[3]) for group in charges_by_group.keys() if group.startswith(f"{age_range}_{sex}_{smoker}"))):
                for group, avg_charges in avg_charges_by_group.items():
                    if group.startswith(f"{age_range}_{sex}_{smoker}_{children}"):
                        print(f" - {sex} {'smokers' if smoker == 'yes' else 'non-smokers'}, with {children} children: {avg_charges}")

Average charges for age range 15-24:
 - female smokers, with 0 children: 24365.79
 - female smokers, with 1 children: 22919.2
 - female smokers, with 2 children: 36976.74
 - female smokers, with 3 children: 26909.52
 - female non-smokers, with 0 children: 3538.66
 - female non-smokers, with 1 children: 2607.38
 - female non-smokers, with 2 children: 9437.64
 - female non-smokers, with 3 children: 7290.92
 - female non-smokers, with 4 children: 4561.19
 - female non-smokers, with 5 children: 4759.21
 - male smokers, with 0 children: 29779.39
 - male smokers, with 1 children: 26571.3
 - male smokers, with 2 children: 30740.41
 - male smokers, with 3 children: 36827.22
 - male smokers, with 4 children: 17942.11
 - male non-smokers, with 0 children: 2574.36
 - male non-smokers, with 1 children: 4184.34
 - male non-smokers, with 2 children: 6493.32
 - male non-smokers, with 3 children: 3663.44
 - male non-smokers, with 4 children: 17128.43
 - male non-smokers, with 5 children: 4915.06
Avera

# <<'FINDINGS'>>

#### BIAS IN DATA:

The Median age (39.0) is very close to the Average age (39.21), thus no age bias in the data.

The ratio between males and females is 1.02 : 1 for males, this is very close, there is no bias in sex in the data.

The Median BMI (30.4) is only 0.22 away from the Average BMI (30.66), thus no BMI bias in the data.

The ratio of people with 1-5 children to people with no children is 1.33 : 1 for people with 1-5 children, there is bias towards people with no children, as I expected a large ratio for the people with 1-5 children, due to the fact that they have a larger range.

The ratio of non-smokers to smokers is 3.88 to 1 for non-smokers, although this is a good sign for health, the data is biased towards non smokers.

The ratio between the regoins of Southwest, Southeast, Northwest, Northeast respectively is 0.24:0.27:0.24:0.24, this is close enough, thus there is no regional bias in the data.

#### DERIVED FROM DATA:

Averages cost increases steadily over age groups in increments on 10 years.

Even though BMI doesn't prove your ultimate health, a steady increase in cost with a increase in BMI is observed.

Costs for males is on average 11.04% higher than that of females.

Amount of children do not play a big roll in the cost.

Smoking greatly effects cost price, since on average smokers pay 280% more than non smokers.

Region does not have a significant effect on the cost price.

On average Male smokers and non-smokers pay more than Female smokers and non-smokers, this can be attributed to the sex factor.

When you look at age, sex, smoking status and children, you observe Males paying more than Females, with a steady increase over the age groups, this indicates, that age, smoking status and amount of children effects both equally, but sex is the determining factor for the difference in cost.

#### Predictions

The Older you are the more you pay.

If you are a Male, you pay more than females.

Smoking will drematically increase your cost price.

The amount of children you have will effect both genders equally, and does not have that big effect on the cost.

The region you live in, does not have a significant effect.

The higher you BMI, the more you pay.